planner/core/planbuilder.go

```diff
@@ -834,6 +834,10 @@ func (b *PlanBuilder) buildSet(ctx context.Context, v *ast.SetStmt) (Plan, error
 			err := ErrSpecificAccessDenied.GenWithStackByArgs("RESTRICTED_VARIABLES_ADMIN")
 			b.visitInfo = appendDynamicVisitInfo(b.visitInfo, "RESTRICTED_VARIABLES_ADMIN", false, err)
 		}
+		if !vars.IsGlobal && vars.Name == "max_allowed_packet" {
```

Comment:
You can add a `Validation` func for `max_allowed_packet`, which is a common way to solve this kind of problem.

Reply:
THKS, I made a new fix follow suggestion, please check. BR

Reply:
Maybe you misunderstand my words. You can add a `Validation` function like this https://github.com/pingcap/tidb/blob/master/sessionctx/variable/sysvar.go#L968

Reply:
Maybe you misunderstand my words. You can add a `Validation` function like this https://github.com/pingcap/tidb/blob/master/sessionctx/variable/sysvar.go#L968

Reply:
THKS, sorry for misunderstand. I did try in sysVar.go to do sth before, but every time when the server started error return with max_allowed_packet. I'll learn and try again.

Reply:
Maybe caused by https://github.com/pingcap/tidb/blob/master/session/session.go#L1214. So there are two ways 
* use some way to figure out it's an intertnal session or not in `Validation` function.
* fix it in planner like this PR.

Both OK for me, do you have some comments? @morgo

Reply:
I fixed it in Validation as suggested in internal session case. please check, THKS much

Reply:
Nice trick to see if it is a user session! LGTM